astropy/nddata/mixins/ndslicing.py

```diff
@@ -4,36 +4,241 @@
+                return self._mask[item]
+            else:
+                log.info("Mask has not the same shape as data.")
         else:
-            new_mask = None
+            log.info("Mask is considered not sliceable.")
+        log.info("Therefore the mask will not be sliced.")
+        return self._mask
 
-        if self.wcs is not None:
-            new_wcs = self.wcs[item]
+    def _slice_wcs(self, item):
+        if self._wcs is None:
+            return None
+        elif isinstance(self._wcs, np.ndarray):
+            if self._wcs.shape == self.data.shape:
+                return self._wcs[item]
+            else:
+                log.info("WCS has not the same shape as data.")
+        elif isinstance(self._wcs, WCS):
+            # There might be a problem if we don't slice all dimensions with
+            # WCS ... e.g. 2D image but slice is [2]
+            # Does it need a shape check as well?
+            return self._wcs[item]
         else:
-            new_wcs = None
+            log.info("WCS is considered not sliceable.")
+        log.info("Therefore the WCS will not be sliced.")
+        return self._wcs
+
+    # Get the documentation in place.
+    if isinstance(_slice.__doc__, string_types):
+        doc = """
+        Controls how the {attr} is sliced.
+
+        Should be called from :meth:`NDSlicingMixin.__getitem__` with ``item``
+        being the slice passed to it.
+
+        Parameters
+        ----------
+        item: Slice
+            The slice passed to :meth:`NDSlicingMixin.__getitem__`.
+
+        Returns
+        -------
+        new_{attr}: same type as ``self._{attr}`` or ``None``
+            The sliced {attr}. If this function doesn't know how to slice
+            the {attr} the return is just the saved ``_{attr}`` and if there is
+            no {attr} the return will default to ``None``.
 
-        return self.__class__(new_data, uncertainty=new_uncertainty,
-                              mask=new_mask, wcs=new_wcs,
-                              meta=self.meta, unit=self.unit)
+        Notes
+        -----
+        This function can treat {attr} if it is
+        {allowed}.
+        The primary purpose of this function is that subclasses using `NDData`
+        with `NDSlicingMixin` with other {attr} restrictions need not to
+        override the complete :meth:`NDSlicingMixin.__getitem__` but only need
+        to override this method.
+        """
+        _slice_uncertainty.__doc__ = doc.format(attr="uncertainty",
+            allowed="`~numpy.ndarray` or `NDUncertainty` with the same shape "
+                    "as the data")
+        _slice_mask.__doc__ = doc.format(attr="mask",
+            allowed="`~numpy.ndarray` with the same shape as the data")
+        _slice_wcs.__doc__ = doc.format(attr="wcs",
+            allowed="`~numpy.ndarray` with the same shape as the data or a "
+                    "`~astropy.wcs.WCS` object")
+        del doc
```

Comment:
No problem, just out of curiosity is there a way to make a decorator keep the function signature and not revert to ``real_function_name(*args, **kwargs)`` because that hides default values and doesn't look so good in the online documentation.